src/spock_recovery.c

```diff
@@ -0,0 +1,489 @@
+	{
+		/* No database name - mark slot as empty */
+		slot->slot_name[0] = '\0';
+	}
+	else
+	{
+		/* Generate and store the slot name */
+		slot_name = get_recovery_slot_name(database_name);
+		if (slot_name)
+		{
+			strncpy(slot->slot_name, slot_name, NAMEDATALEN - 1);
+			slot->slot_name[NAMEDATALEN - 1] = '\0';
+			pfree(slot_name);
+		}
+		else
+		{
+			slot->slot_name[0] = '\0';
+		}
+	}
+
+	/* Initialize all tracking fields to empty/invalid state */
+	slot->restart_lsn = InvalidXLogRecPtr;
+	slot->confirmed_flush_lsn = InvalidXLogRecPtr;
+	slot->min_unacknowledged_ts = 0;
+	slot->active = false;
+	slot->in_recovery = false;
+
+	/* Increment generation counter (atomic operation, no lock needed) */
+	pg_atomic_add_fetch_u32(&slot->recovery_generation, 1);
+}
+
+/*
+ * create_recovery_slot
+ *
+ * Create an inactive logical replication slot for catastrophic failure recovery.
+ *
+ * Creates a persistent replication slot that preserves WAL segments for all
+ * subscriptions in the specified database. The slot remains inactive and is
+ * never used for normal replication.
+ *
+ * If a recovery slot already exists for this database, returns success without
+ * creating a duplicate. This makes the function safe to call multiple times.
+ *
+ * Args:
+ *		database_name: Name of the database to create recovery slot for
+ *
+ * Returns:
+ *		true if slot exists (created now or previously)
+ *		false on error (logged as WARNING, not fatal)
+ */
+bool
+create_recovery_slot(const char *database_name)
+{
+	SpockRecoverySlotData *slot;
+	char	   *slot_name;
+	bool		success = false;
+	bool		cleanup_needed = false;
+
+	if (!SpockRecoveryCtx)
+	{
+		elog(WARNING, "recovery coordinator not initialized");
+		return false;
+	}
+
+	if (!database_name)
+	{
+		elog(WARNING, "database name cannot be NULL");
+		return false;
+	}
+
+	LWLockAcquire(SpockRecoveryCtx->lock, LW_EXCLUSIVE);
+
+	slot = &SpockRecoveryCtx->recovery_slot;
+
+	/* If slot already exists, we're done */
+	if (slot->active)
+	{
+		LWLockRelease(SpockRecoveryCtx->lock);
+		elog(LOG, "recovery slot already exists");
+		return true;
+	}
+
+	/* Initialize the slot metadata */
+	initialize_recovery_slot(slot, database_name);
+
+	/* Copy slot name for use outside the lock */
+	slot_name = pstrdup(slot->slot_name);
+	LWLockRelease(SpockRecoveryCtx->lock);
+
+	/*
+	 * Create the actual replication slot
+	 *
+	 * Parameters:
+	 * - slot_name: name of the slot to create
+	 * - true: database-specific slot
+	 * - RS_PERSISTENT: slot survives server restart
+	 * - false: not two-phase (we don't need prepared transaction support)
+	 * - false: not failover (recovery slots don't need failover)
+	 * - false: not synced (recovery slots are local only)
+	 */
+	PG_TRY();
+	{
+		ReplicationSlotCreate(slot_name, true, RS_PERSISTENT, false, false, false);
+		success = true;
+
+		elog(LOG, "created recovery slot '%s' (INACTIVE - for catastrophic failure recovery only)",
+			 slot_name);
+	}
+	PG_CATCH();
+	{
```

Comment:
Let’s fail on error here. If recovery slots are enabled, failure to create one should be treated as a fatal error and result in an exit.

Using PG_TRY/CATCH just to swallow the failure complicates things unnecessarily; we can drop that and keep the logic simple.

Also, only touch shared-memory state after a successful create; that removes the success / cleanup_needed flags and the reset path entirely.